mdtraj/formats/pdbx/pdbxfile.py

```diff
@@ -0,0 +1,727 @@
+Simbios, the NIH National Center for Physics-Based Simulation of
+Biological Structures at Stanford, funded under the NIH Roadmap for
+Medical Research, grant U54 GM072970. See https://simtk.org.
+
+Portions copyright (c) 2015-2025 Stanford University and the Authors.
+Authors: Peter Eastman
+Contributors: Jason Swails
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS, CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+__author__ = "Peter Eastman"
+__version__ = "2.0"
+
+import sys
+import warnings
+from collections import defaultdict
+from datetime import date
+
+import numpy as np
+
+import mdtraj._version
+from mdtraj.core import element as elem
+from mdtraj.core import topology
+from mdtraj.core.topology import Topology
+from mdtraj.utils import (
+    open_maybe_zipped,
+)
+
+from .PdbxReader import PdbxReader
+
+
+class PDBxFile:
+    """PDBxFile parses a PDBx/mmCIF file and constructs a Topology and a set of atom positions from it."""
+
+    def __init__(self, file):
+        """Load a PDBx/mmCIF file.
+
+        The atom positions and Topology can be retrieved by calling
+        getPositions() and getTopology().
+
+        Parameters
+        ----------
+        file : string
+            the name of the file to load.  Alternatively you can pass an open
+            file object.
+        """
+        top = Topology()
+        ## The Topology read from the PDBx/mmCIF file
+        self.topology = top
+        self._positions = []
+        self._unitcell_lengths = None
+        self._unitcell_angles = None
+
+        self._loadNameReplacementTables()
+
+        # Load the file.
+
+        inputFile = file
+        ownHandle = False
+        if isinstance(file, str):
+            inputFile = open_maybe_zipped(file, "r")
+            ownHandle = True
+        reader = PdbxReader(inputFile)
+        data = []
+        reader.read(data)
+        if ownHandle:
+            inputFile.close()
+        block = data[0]
+
+        # Build the topology.
+
+        atomData = block.getObj("atom_site")
```

Comment:
I find this section really wordy with a lot of repeating calls (and loads of variables), only later to realize it's all to find the column numbers for each dataset. Maybe it's easier to do this in a dictionary instead? It would also be nice to have a comment saying `# Mapping datasets to respective column numbers` or something